clap_derive/src/derives/args.rs

```diff
@@ -51,10 +52,169 @@ pub(crate) fn derive_args(input: &DeriveInput) -> Result<TokenStream, syn::Error
                 .collect::<Result<Vec<_>, syn::Error>>()?;
             gen_for_struct(&item, ident, &input.generics, &fields)
         }
+        Data::Enum(DataEnum { ref variants, .. }) => {
+            let name = Name::Derived(ident.clone());
+            let item = Item::from_args_struct(input, name)?;
+
+            let variant_items = variants
+                .iter()
+                .map(|variant| {
+                    let item = Item::from_args_enum_variant(variant)?;
+                    Ok((item, variant))
+                })
+                .collect::<Result<Vec<_>, syn::Error>>()?;
+
+            gen_for_enum(&item, ident, &input.generics, &variant_items)
+        }
         _ => abort_call_site!("`#[derive(Args)]` only supports non-tuple structs"),
     }
 }
 
+pub(crate) fn gen_for_enum(
+    _item: &Item,
+    item_name: &Ident,
+    generics: &Generics,
+    variants: &[(Item, &Variant)],
+) -> Result<TokenStream, syn::Error> {
+    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
+    let app_var = Ident::new("__clap_app", Span::call_site());
+    let mut augmentations = TokenStream::default();
+    let mut augmentations_update = TokenStream::default();
+
+    let mut constructors = TokenStream::default();
+
+    for (item, variant) in variants.iter() {
+        let Fields::Named(ref fields) = variant.fields else {
+            abort! { variant.span(),
+                "`#[derive(Args)]` only supports named enum variants if used on an enum",
+            }
+        };
```

Comment:
Restricting this to named enums only, allows to reuse the struct codegen with little modification.

`Unnamed` variants, with a single item might somehow dispatch the parsing bits to the contained Type and _somehow_ set the group conflicts using `ArgGroup::mut_group` on the augment side and ??? on the from arg matches (that might simply forward?).

* `Unit` variants need to be parsed as flags?
* What about `enum Something { Variant(String) }` would we expect this to parse as 1) a positional, 2) a flag, 3) not at all bc we cant forward to `String`s implementations?

Reply:
I'm fine with scoping the first PR toNamed variants.  We should have compile error tests for all of the other kinds of variants.

I could see soon after support single-element tuple variants with `#[command(flatten)] Variant(MoreArgs)` to avoid the need for `Variant { #[command(flatten)] args: MoreArgs }`.  I'm assuming supporting specifically that wouldn't be too bad.

Longer term...

`Unit` variant should probably be discussed in the issue.  My first guess is to use it as an "else" case.

For other single-element tuples, I see it working just like a field.  They are positional by default and need `#[arg(short, long)]` on the variant to change it to something else.  We'd use `value_parser!` to understand what to do with the type inside of the variant.

For N-element tuples, we have an issue for supporting those on fields and I'd point to that for variants as well.

Reply:
that all makes sense in my book.
I'm still getting familiar with `Item` and the parsing around it where Named variants were just the closest to intuitively scrape together.
Happy to look into that soon after.